kernels/volk/volk_32fc_index_max_16u.h

```diff
@@ -141,104 +138,33 @@ volk_32fc_index_max_16u_a_avx2(uint16_t* target, lv_32fc_t* src0,
-
-      src0 += 2;
-
-      xmm1 = _mm256_hadd_ps(xmm2, xmm2);
-
-      xmm3 = _mm256_max_ps(xmm1, xmm3);
-
-      xmm4.float_vec = _mm256_cmp_ps(xmm1, xmm3, _CMP_LT_OS);
-      xmm5.float_vec = _mm256_cmp_ps(xmm1, xmm3, _CMP_EQ_OQ);
-
-      xmm11 = _mm256_and_si256(xmm8, xmm5.int_vec);
-      xmm12 = _mm256_and_si256(xmm9, xmm4.int_vec);
-
-      xmm9 = _mm256_add_epi32(xmm11, xmm12);
-
-      xmm8 = _mm256_add_epi32(xmm8, xmm10);
-  }
-
-  /*
-  idx = _mm256_setzero_si256();
-  for(i = 0; i < leftovers2; ++i) {
-    //printf("%u, %u, %u, %u\n", ((uint32_t*)&xmm9)[0], ((uint32_t*)&xmm9)[1], ((uint32_t*)&xmm9)[2], ((uint32_t*)&xmm9)[3]);
-
-    sq_dist = lv_creal(src0[0]) * lv_creal(src0[0]) + lv_cimag(src0[0]) * lv_cimag(src0[0]);
-
-    //xmm = _mm_load1_ps(&sq_dist);//insert?
-    xmm2 = _mm256_set1_ps(sq_dist);
-    //xmm2 = _mm256_insertf128_ps(xmm2, xmm, 0);
-
-    xmm1 = xmm3;
-
-    xmm3 = _mm256_max_ps(xmm3, xmm2);//only lowest 32bit value
-    xmm3 = _mm256_permutevar8x32_ps(xmm3, idx);
-
-    xmm4.float_vec = _mm256_cmp_ps(xmm1, xmm3, _CMP_LT_OS);
-    xmm5.float_vec = _mm256_cmp_ps(xmm1, xmm3, _CMP_EQ_OQ);
-
-    xmm8 = _mm256_permutevar8x32_epi32(xmm8, idx);
-
-    xmm11 = _mm256_and_si256(xmm8, xmm4.int_vec);
-    xmm12 = _mm256_and_si256(xmm9, xmm5.int_vec);
-
-    xmm9 = _mm256_add_epi32(xmm11, xmm12);
-}*/
 
   _mm256_store_ps((float*)&(holderf.f), xmm3);
   _mm256_store_si256(&(holderi.int_vec), xmm9);
 
-  target[0] = holderi.i[0];
-  sq_dist = holderf.f[0];
-  target[0] = (holderf.f[1] > sq_dist) ? holderi.i[1] : target[0];
-  sq_dist = (holderf.f[1] > sq_dist) ? holderf.f[1] : sq_dist;
-  target[0] = (holderf.f[2] > sq_dist) ? holderi.i[2] : target[0];
-  sq_dist = (holderf.f[2] > sq_dist) ? holderf.f[2] : sq_dist;
-  target[0] = (holderf.f[3] > sq_dist) ? holderi.i[3] : target[0];
-  sq_dist = (holderf.f[3] > sq_dist) ? holderf.f[3] : sq_dist;
-  target[0] = (holderf.f[4] > sq_dist) ? holderi.i[4] : target[0];
-  sq_dist = (holderf.f[4] > sq_dist) ? holderf.f[4] : sq_dist;
-  target[0] = (holderf.f[5] > sq_dist) ? holderi.i[5] : target[0];
-  sq_dist = (holderf.f[5] > sq_dist) ? holderf.f[5] : sq_dist;
-  target[0] = (holderf.f[6] > sq_dist) ? holderi.i[6] : target[0];
-  sq_dist = (holderf.f[6] > sq_dist) ? holderf.f[6] : sq_dist;
-  target[0] = (holderf.f[7] > sq_dist) ? holderi.i[7] : target[0];
-  sq_dist = (holderf.f[7] > sq_dist) ? holderf.f[7] : sq_dist;
+  for(i = 0; i < 8; i++)
+  {
+    if(holderf.f[i] > sq_dist)
+    {
+      target[0] = holderi.i[i];
+      sq_dist = holderf.f[i];
+    }
+    else if(holderf.f[i] == sq_dist)
+    {
+      if (target[0] > holderi.i[i])
+        target[0] = holderi.i[i];
+    }
+  }
 
+  for(i = bound * 8; i < num_points; i++, src0++)
+  {
+    float new_sq_dist = lv_creal(src0[0]) * lv_creal(src0[0]) + lv_cimag(src0[0]) * lv_cimag(src0[0]);
+    if(new_sq_dist > sq_dist)
+    {
+      target[0] = i;
+      sq_dist = new_sq_dist;
+    }
+  }
```

Comment:
> Could you just call the generic kernel on the tail instead?

Not easily, because the kernel only returns the index and not the maximum value itself (which would then need to be recomputed to check whether it's larger than the beginning of the array. That's why I went with the copy & paste approach. The code is actually adapted from the `32f_index_max` AVX kernels, but I'll see if I can make it more like the `32fc_index_max` generic kernels.

Reply:
You could write a function that also returns the maximum value. Then just call that function in the generic kernel.